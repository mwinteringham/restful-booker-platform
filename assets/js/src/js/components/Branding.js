import React, { useEffect, useState } from 'react';
import { API } from '../libs/Api.js';
import ReactModal from 'react-modal';

const Branding = () => {
    
    const [branding, setBranding] = useState({
        map: {
            latitude: 0,
            longitude: 0
        },
        logoUrl: '',
        description: '',
        name: '',
        contact: {
            name: '',
            address: '',
            phone: '',
            email: ''
        }
    })
    const [showModal, toggleModal] = useState(false);
    const [errors, setErrors] = useState([]);

    useEffect(() => {
        API.getBranding(setBranding);
    }, [])

    const doUpdate = () => {
        API.putBranding(branding, setErrors, toggleModal);
    }

    const updateState = (event) => {
        switch(event.id){
            case 'latitude':
                branding.map.latitude = event.value;
                break;
            case 'longitude':
                branding.map.longitude = event.value;
                break;
            case 'contactName':
                branding.contact.name = event.value;
                break;
            case 'contactAddress':
                branding.contact.address = event.value;
                break;
            case 'contactPhone':
                branding.contact.phone = event.value;
                break;
            case 'contactEmail':
                branding.contact.email = event.value;
                break;
            default :
                branding[event.id] = event.value;
                break;
        }

        setBranding(prevState => ({
            ...prevState
        }));
    }


    let errorMessages = '';
    
    if(errors.length > 0){
        errorMessages = <div className="alert alert-danger" style={{marginTop : 1 + "rem"}}>
            {errors.map((value) => {
                return <p key={value}>{value}</p>
            })}
        </div>
    }

    return <div className="branding-form">
                <h2>B&amp;B details</h2>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Name</span>
                    </div>
                    <input type="text" className="form-control" id="name" value={branding.name} onChange={e => updateState(e.target)} placeholder="Enter B&amp;B name" />
                </div>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Logo</span>
                    </div>
                    <input type="text" className="form-control" id="logoUrl" value={branding.logoUrl} onChange={e => updateState(e.target)} placeholder="Enter image url" />
                </div>
                <div className="input-group">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Description</span>
                    </div>
                    <textarea className="form-control" value={branding.description} onChange={e => updateState(e.target)} id="description" rows="5"></textarea>
                </div>
                <br />
                <h2>Map details</h2>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Latitude</span>
                    </div>
                    <input type="text" className="form-control" id="latitude" value={branding.map.latitude} onChange={e => updateState(e.target)} placeholder="Enter Latitude" />
                </div>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Longitude</span>
                    </div>
                    <input type="text" className="form-control" id="longitude" value={branding.map.longitude} onChange={e => updateState(e.target)} placeholder="Enter Longitude" />
                </div>
                <br />
                <h2>Contact details</h2>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Name</span>
                    </div>
                    <input type="text" className="form-control" id="contactName" value={branding.contact.name} onChange={e => updateState(e.target)} placeholder="Enter Contact Name" />
                </div>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Address</span>
                    </div>
                    <input type="text" className="form-control" id="contactAddress" value={branding.contact.address} onChange={e => updateState(e.target)} placeholder="Enter Address" />
                </div>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Phone</span>
                    </div>
                    <input type="text" className="form-control" id="contactPhone" value={branding.contact.phone} onChange={e => updateState(e.target)} placeholder="Enter Phone Number" />
                </div>
                <div className="input-group mb-3">
                    <div className="input-group-prepend">
                        <span className="input-group-text">Email</span>
                    </div>
                    <input type="email" className="form-control" id="contactEmail" value={branding.contact.email} onChange={e => updateState(e.target)} placeholder="Enter Email Address" />
                </div>
                <button type="submit" id="updateBranding" className="btn btn-outline-primary" onClick={doUpdate}>Submit</button>
                <ReactModal 
                    isOpen={showModal}
                    contentLabel="onRequestClose Example"
                    onRequestClose={() => {toggleModal(false)}}
                    className="Modal"
                    >
                    
                    <div className="form-row text-center">
                        <div className="col-12">
                            <p>Branding updated!</p>
                            <button className="btn btn-outline-primary" onClick={() => {toggleModal(false)}}>Close</button>
                        </div>
                    </div>
                </ReactModal>
                {errorMessages}
            </div>

}

export default Branding;
