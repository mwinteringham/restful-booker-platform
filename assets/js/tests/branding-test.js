import React from 'react';
import Branding from '../src/js/components/Branding.js';
import nock from 'nock';
import ReactModal from 'react-modal';
import '@testing-library/jest-dom'
import {
    render,
    fireEvent,
    waitFor
  } from '@testing-library/react'

const brandingData = {
    name: 'Shady Meadows B&B',
    map: {
        latitude: 52.6351204,
        longitude: 1.2733774
    },
    logoUrl: 'https://www.mwtestconsultancy.co.uk/img/rbp-logo.png',
    description: 'Welcome to Shady Meadows, a delightful Bed & Breakfast nestled in the hills on Newingtonfordburyshire. A place so beautiful you will never want to leave. All our rooms have comfortable beds and we provide breakfast from the locally sourced supermarket. It is a delightful place.',
    contact: {
        name: 'Shady Meadows B&B',
        address: 'The Old Farmhouse, Shady Street, Newfordburyshire, NE1 410S',
        phone: '0123456789',
        email: 'fake@fakeemail.com'
    }
}

nock('http://localhost')
    .persist()
    .get('/branding/')
    .reply(200, brandingData)

test('Branding page renders', async () => {
    const { asFragment, findByDisplayValue } = render(
        <Branding />
    )

    await findByDisplayValue("52.6351204")
    
    expect(asFragment()).toMatchSnapshot();
});

test('Branding page shows modal on success', async () => {
    nock('http://localhost')
        .put('/branding/')
        .reply(202)

    ReactModal.setAppElement(document.createElement('div'));

    const {getByText, getByPlaceholderText} = render(
        <Branding />
    )

    fireEvent.change(getByPlaceholderText('Enter B&B name'), { target: { value: 'Updated Room' } });
    fireEvent.click(getByText('Submit'))

    await waitFor(() => expect(getByText('Branding updated!')).toBeInTheDocument())
});

test('Branding page shows errors', async () => {
    nock('http://localhost')
        .put('/branding/')
        .reply(400, {
            "fieldErrors": ["Phone should not be blank"]
        })

    const {getByText, findByText} = render(
        <Branding />
    )

    fireEvent.click(getByText('Submit'))
    await findByText('Phone should not be blank')

    expect(getByText('Phone should not be blank')).toBeInTheDocument()
});
