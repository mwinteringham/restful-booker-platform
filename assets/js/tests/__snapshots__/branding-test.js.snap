// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Branding page renders 1`] = `
<DocumentFragment>
  <div
    class="branding-form"
  >
    <h2>
      B&B details
    </h2>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Name
        </span>
      </div>
      <input
        class="form-control"
        id="name"
        placeholder="Enter B&B name"
        type="text"
        value="Shady Meadows B&B"
      />
    </div>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Logo
        </span>
      </div>
      <input
        class="form-control"
        id="logoUrl"
        placeholder="Enter image url"
        type="text"
        value="https://www.mwtestconsultancy.co.uk/img/rbp-logo.png"
      />
    </div>
    <div
      class="input-group"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Description
        </span>
      </div>
      <textarea
        class="form-control"
        id="description"
        rows="5"
      >
        Welcome to Shady Meadows, a delightful Bed & Breakfast nestled in the hills on Newingtonfordburyshire. A place so beautiful you will never want to leave. All our rooms have comfortable beds and we provide breakfast from the locally sourced supermarket. It is a delightful place.
      </textarea>
    </div>
    <br />
    <h2>
      Map details
    </h2>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Latitude
        </span>
      </div>
      <input
        class="form-control"
        id="latitude"
        placeholder="Enter Latitude"
        type="text"
        value="52.6351204"
      />
    </div>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Longitude
        </span>
      </div>
      <input
        class="form-control"
        id="longitude"
        placeholder="Enter Longitude"
        type="text"
        value="1.2733774"
      />
    </div>
    <br />
    <h2>
      Contact details
    </h2>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Name
        </span>
      </div>
      <input
        class="form-control"
        id="contactName"
        placeholder="Enter Contact Name"
        type="text"
        value="Shady Meadows B&B"
      />
    </div>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Address
        </span>
      </div>
      <input
        class="form-control"
        id="contactAddress"
        placeholder="Enter Address"
        type="text"
        value="The Old Farmhouse, Shady Street, Newfordburyshire, NE1 410S"
      />
    </div>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Phone
        </span>
      </div>
      <input
        class="form-control"
        id="contactPhone"
        placeholder="Enter Phone Number"
        type="text"
        value="0123456789"
      />
    </div>
    <div
      class="input-group mb-3"
    >
      <div
        class="input-group-prepend"
      >
        <span
          class="input-group-text"
        >
          Email
        </span>
      </div>
      <input
        class="form-control"
        id="contactEmail"
        placeholder="Enter Email Address"
        type="email"
        value="fake@fakeemail.com"
      />
    </div>
    <button
      class="btn btn-outline-primary"
      id="updateBranding"
      type="submit"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
